{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Prepare data tables for enhancer counting and other information that will be used to generate figures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>obs_score</th>\n",
       "      <th>p63RE_class</th>\n",
       "      <th>strand</th>\n",
       "      <th>p63RE_chrom</th>\n",
       "      <th>p63RE_start</th>\n",
       "      <th>p63RE_stop</th>\n",
       "      <th>enhancer_variant</th>\n",
       "      <th>seq</th>\n",
       "      <th>id</th>\n",
       "      <th>loc</th>\n",
       "      <th>p63RE_type</th>\n",
       "      <th>gc</th>\n",
       "      <th>Number of supporting p53 ChIP-seq datasets_p53</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>63639</th>\n",
       "      <td>9</td>\n",
       "      <td>primary</td>\n",
       "      <td>+</td>\n",
       "      <td>chr11</td>\n",
       "      <td>79043019</td>\n",
       "      <td>79043038</td>\n",
       "      <td>flankShuffle</td>\n",
       "      <td>ttggccaaattgcgacccacctgttgattgccccctcacagaaacc...</td>\n",
       "      <td>chr11_79043019_79043038_flankShuffle</td>\n",
       "      <td>chr11_79043019_79043038</td>\n",
       "      <td>p53RE+p63RE</td>\n",
       "      <td>52.941176</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3060</th>\n",
       "      <td>16</td>\n",
       "      <td>primary</td>\n",
       "      <td>-</td>\n",
       "      <td>chr8</td>\n",
       "      <td>143385410</td>\n",
       "      <td>143385434</td>\n",
       "      <td>WT</td>\n",
       "      <td>cggggggcacagctgccccgaggggccggctcaaggctgtccctgc...</td>\n",
       "      <td>chr8_143385410_143385434_WT</td>\n",
       "      <td>chr8_143385410_143385434</td>\n",
       "      <td>Unique p63RE</td>\n",
       "      <td>73.333333</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80692</th>\n",
       "      <td>9</td>\n",
       "      <td>primary</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1</td>\n",
       "      <td>191901820</td>\n",
       "      <td>191901839</td>\n",
       "      <td>fullShuffle</td>\n",
       "      <td>tgaggacgagcctgcagttgtttactagtgcctacgcttagtgaac...</td>\n",
       "      <td>chr1_191901820_191901839_fullShuffle</td>\n",
       "      <td>chr1_191901820_191901839</td>\n",
       "      <td>p53RE+p63RE</td>\n",
       "      <td>47.899160</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74645</th>\n",
       "      <td>13</td>\n",
       "      <td>secondary</td>\n",
       "      <td>-</td>\n",
       "      <td>chr11</td>\n",
       "      <td>35894092</td>\n",
       "      <td>35894106</td>\n",
       "      <td>fullShuffle</td>\n",
       "      <td>aggcccttcgagctataagagtctgaaaatacgtcatttcatgtat...</td>\n",
       "      <td>chr11_35894092_35894106_fullShuffle</td>\n",
       "      <td>chr11_35894092_35894106</td>\n",
       "      <td>p53RE+p63RE</td>\n",
       "      <td>54.166667</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29799</th>\n",
       "      <td>9</td>\n",
       "      <td>primary</td>\n",
       "      <td>-</td>\n",
       "      <td>chr2</td>\n",
       "      <td>68368505</td>\n",
       "      <td>68368529</td>\n",
       "      <td>mut</td>\n",
       "      <td>taacagcaaagcctaccatctcattgatcccaggcagcattctgaa...</td>\n",
       "      <td>chr2_68368505_68368529_mut</td>\n",
       "      <td>chr2_68368505_68368529</td>\n",
       "      <td>Unique p63RE</td>\n",
       "      <td>50.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11510</th>\n",
       "      <td>9</td>\n",
       "      <td>quinary</td>\n",
       "      <td>+</td>\n",
       "      <td>chr10</td>\n",
       "      <td>5564328</td>\n",
       "      <td>5564337</td>\n",
       "      <td>WT</td>\n",
       "      <td>cacaagtcagggtgttgcctctgcttacaagcctttggtgactgcc...</td>\n",
       "      <td>chr10_5564328_5564337_WT</td>\n",
       "      <td>chr10_5564328_5564337</td>\n",
       "      <td>p53RE+p63RE</td>\n",
       "      <td>60.504202</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27148</th>\n",
       "      <td>10</td>\n",
       "      <td>tertiary</td>\n",
       "      <td>+</td>\n",
       "      <td>chr11</td>\n",
       "      <td>119229872</td>\n",
       "      <td>119229886</td>\n",
       "      <td>mut</td>\n",
       "      <td>tgggttcaagtgattctcctgcctcagcctcccgagtagctggaat...</td>\n",
       "      <td>chr11_119229872_119229886_mut</td>\n",
       "      <td>chr11_119229872_119229886</td>\n",
       "      <td>Unique p63RE</td>\n",
       "      <td>43.333333</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46853</th>\n",
       "      <td>9</td>\n",
       "      <td>secondary</td>\n",
       "      <td>+</td>\n",
       "      <td>chr17</td>\n",
       "      <td>31492027</td>\n",
       "      <td>31492041</td>\n",
       "      <td>shuffle</td>\n",
       "      <td>aaaccagcaaagggcacgccggcgatgtgcatacagctgtgtctct...</td>\n",
       "      <td>chr17_31492027_31492041_shuffle</td>\n",
       "      <td>chr17_31492027_31492041</td>\n",
       "      <td>p53RE+p63RE</td>\n",
       "      <td>62.500000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23606</th>\n",
       "      <td>13</td>\n",
       "      <td>primary</td>\n",
       "      <td>-</td>\n",
       "      <td>chr6</td>\n",
       "      <td>157260962</td>\n",
       "      <td>157260981</td>\n",
       "      <td>mut</td>\n",
       "      <td>tgcatgattagacttgcttgatcttcccctctgcccttgccaggag...</td>\n",
       "      <td>chr6_157260962_157260981_mut</td>\n",
       "      <td>chr6_157260962_157260981</td>\n",
       "      <td>p53RE+p63RE</td>\n",
       "      <td>56.302521</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71417</th>\n",
       "      <td>17</td>\n",
       "      <td>primary</td>\n",
       "      <td>+</td>\n",
       "      <td>chr7</td>\n",
       "      <td>99561554</td>\n",
       "      <td>99561573</td>\n",
       "      <td>fullShuffle</td>\n",
       "      <td>ctgtgcgtgacaaccaacagtatctaataagtttacgggccatatg...</td>\n",
       "      <td>chr7_99561554_99561573_fullShuffle</td>\n",
       "      <td>chr7_99561554_99561573</td>\n",
       "      <td>p53RE+p63RE</td>\n",
       "      <td>45.378151</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       obs_score p63RE_class strand p63RE_chrom  p63RE_start  p63RE_stop  \\\n",
       "63639          9     primary      +       chr11     79043019    79043038   \n",
       "3060          16     primary      -        chr8    143385410   143385434   \n",
       "80692          9     primary      -        chr1    191901820   191901839   \n",
       "74645         13   secondary      -       chr11     35894092    35894106   \n",
       "29799          9     primary      -        chr2     68368505    68368529   \n",
       "11510          9     quinary      +       chr10      5564328     5564337   \n",
       "27148         10    tertiary      +       chr11    119229872   119229886   \n",
       "46853          9   secondary      +       chr17     31492027    31492041   \n",
       "23606         13     primary      -        chr6    157260962   157260981   \n",
       "71417         17     primary      +        chr7     99561554    99561573   \n",
       "\n",
       "      enhancer_variant                                                seq  \\\n",
       "63639     flankShuffle  ttggccaaattgcgacccacctgttgattgccccctcacagaaacc...   \n",
       "3060                WT  cggggggcacagctgccccgaggggccggctcaaggctgtccctgc...   \n",
       "80692      fullShuffle  tgaggacgagcctgcagttgtttactagtgcctacgcttagtgaac...   \n",
       "74645      fullShuffle  aggcccttcgagctataagagtctgaaaatacgtcatttcatgtat...   \n",
       "29799              mut  taacagcaaagcctaccatctcattgatcccaggcagcattctgaa...   \n",
       "11510               WT  cacaagtcagggtgttgcctctgcttacaagcctttggtgactgcc...   \n",
       "27148              mut  tgggttcaagtgattctcctgcctcagcctcccgagtagctggaat...   \n",
       "46853          shuffle  aaaccagcaaagggcacgccggcgatgtgcatacagctgtgtctct...   \n",
       "23606              mut  tgcatgattagacttgcttgatcttcccctctgcccttgccaggag...   \n",
       "71417      fullShuffle  ctgtgcgtgacaaccaacagtatctaataagtttacgggccatatg...   \n",
       "\n",
       "                                         id                        loc  \\\n",
       "63639  chr11_79043019_79043038_flankShuffle    chr11_79043019_79043038   \n",
       "3060            chr8_143385410_143385434_WT   chr8_143385410_143385434   \n",
       "80692  chr1_191901820_191901839_fullShuffle   chr1_191901820_191901839   \n",
       "74645   chr11_35894092_35894106_fullShuffle    chr11_35894092_35894106   \n",
       "29799            chr2_68368505_68368529_mut     chr2_68368505_68368529   \n",
       "11510              chr10_5564328_5564337_WT      chr10_5564328_5564337   \n",
       "27148         chr11_119229872_119229886_mut  chr11_119229872_119229886   \n",
       "46853       chr17_31492027_31492041_shuffle    chr17_31492027_31492041   \n",
       "23606          chr6_157260962_157260981_mut   chr6_157260962_157260981   \n",
       "71417    chr7_99561554_99561573_fullShuffle     chr7_99561554_99561573   \n",
       "\n",
       "         p63RE_type         gc  Number of supporting p53 ChIP-seq datasets_p53  \n",
       "63639   p53RE+p63RE  52.941176                                             NaN  \n",
       "3060   Unique p63RE  73.333333                                             NaN  \n",
       "80692   p53RE+p63RE  47.899160                                            15.0  \n",
       "74645   p53RE+p63RE  54.166667                                             NaN  \n",
       "29799  Unique p63RE  50.000000                                             NaN  \n",
       "11510   p53RE+p63RE  60.504202                                             5.0  \n",
       "27148  Unique p63RE  43.333333                                             NaN  \n",
       "46853   p53RE+p63RE  62.500000                                             NaN  \n",
       "23606   p53RE+p63RE  56.302521                                             NaN  \n",
       "71417   p53RE+p63RE  45.378151                                             7.0  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#import libraries\n",
    "import pandas as pd\n",
    "import os\n",
    "import bioframe as bf\n",
    "#import key tables\n",
    "os.chdir(r'Z:\\gbaniulyte\\MPRA\\AM_STARRSeq\\final_scripts_fastq\\tables')\n",
    "#this is the original table with seq information\n",
    "df = pd.read_csv('p63enh_starrseq_info.csv', usecols=['uniqueID', 'obs_score','p63RE_class',\n",
    "\t\t\t\t\t\t\t\t\t\t\t\t\t  'WT', 'mut', 'shuffle','flankShuffle','fullShuffle',\n",
    "\t\t\t\t\t\t\t\t\t\t\t\t\t  'strand','p63RE_chrom','p63RE_start','p63RE_stop'])\n",
    "#make a long table and reformat\n",
    "df_long = df.melt(id_vars=['uniqueID', 'obs_score','p63RE_class','strand','p63RE_chrom','p63RE_start','p63RE_stop'],\n",
    "\t\t\t\t  value_vars=['WT', 'mut', 'shuffle','flankShuffle', 'fullShuffle'],\n",
    "\t\t\t\t  var_name='enhancer_variant',\n",
    "\t\t\t\t  value_name='seq')\n",
    "#add unique id with p63RE locationa and variant\n",
    "df_long['id'] = df_long.apply(lambda row: row.uniqueID[: row.uniqueID.rfind('_') + 1] + row.enhancer_variant, axis='columns')\n",
    "df_long['loc'] = df_long.apply(lambda row: row.uniqueID[: row.uniqueID.rfind('_')], axis='columns')\n",
    "df_long[['p63RE_class', 'p63RE_type']] = df_long.apply(lambda row: [row.p63RE_class[:row.p63RE_class.find('.')],\n",
    "\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t\trow.p63RE_class[row.p63RE_class.find('.') + 1:]],\n",
    "\t\t\t\t\t\t\t\t\t\t\t\t\t    axis='columns', result_type='expand')\n",
    "df_long.replace({'unique-p63RE':'Unique p63RE', 'p53+p63RE' : 'p53RE+p63RE'}, inplace=True)\n",
    "df_long.drop(columns=['uniqueID', ], inplace=True)\n",
    "#add enhancner GC content (%)\n",
    "def gc_count(sequence):\n",
    "\tsequence = sequence.upper()\n",
    "\tgc = ((sequence.count('G') + sequence.count('C')) / len(sequence)) * 100\n",
    "\treturn gc\n",
    "df_long['gc'] = df_long.seq.apply(lambda x: gc_count(x))\n",
    "riege_xls = pd.ExcelFile('elife-63266-supp3-v2.xlsx', engine='openpyxl')\n",
    "p53_motifs = riege_xls.parse('p53peaks+REs', usecols=['Number of supporting p53 ChIP-seq datasets','p53RE_Chromosome','p53RE_Start','p53RE_End'])\n",
    "#remove rows with no motif coordinates\n",
    "p53_motifs = p53_motifs[p53_motifs.p53RE_Start != '-']\n",
    "#bioframe needs int\n",
    "p53_motifs =p53_motifs.astype({\"p53RE_Start\": int, \"p53RE_End\": int})\n",
    "p53_p63_overlap = bf.overlap(df_long, p53_motifs,  cols1 = ['p63RE_chrom','p63RE_start','p63RE_stop'],\n",
    "\t\t\t\t\t\t\t  cols2 = ['p53RE_Chromosome','p53RE_Start','p53RE_End'], suffixes = ['','_p53'])\n",
    "del(riege_xls,p53_motifs)\n",
    "#save enh id that has p53 ob if they overlap with p63RE\n",
    "df_long = df_long.merge(p53_p63_overlap[['id','Number of supporting p53 ChIP-seq datasets_p53']], on='id', how='left')\n",
    "df_long.to_csv('p63enh_starrseq_info_long.csv')\n",
    "df_long.sample(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Prepare raw counts table from all p63 enhancer STARRseq experiments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq</th>\n",
       "      <th>id</th>\n",
       "      <th>loc</th>\n",
       "      <th>enhancer_variant</th>\n",
       "      <th>HaCaT_1_count</th>\n",
       "      <th>HaCaT_2_count</th>\n",
       "      <th>MCF10AGus_count</th>\n",
       "      <th>MCF10Ap53KO_1_count</th>\n",
       "      <th>MCF10Ap53KO_2_count</th>\n",
       "      <th>MCF10ATAp63B_count</th>\n",
       "      <th>MCF10A_1_count</th>\n",
       "      <th>MCF10A_2_count</th>\n",
       "      <th>plasmid_1_count</th>\n",
       "      <th>plasmid_2_count</th>\n",
       "      <th>SCC25_1_count</th>\n",
       "      <th>SCC25_2_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>69090</th>\n",
       "      <td>ACGAGTCCGCTGAAGACCGTCACAACTTCCGTCATCCGTCCGCCTA...</td>\n",
       "      <td>chr9_13027983_13028007_flankShuffle</td>\n",
       "      <td>chr9_13027983_13028007</td>\n",
       "      <td>flankShuffle</td>\n",
       "      <td>559</td>\n",
       "      <td>958</td>\n",
       "      <td>927</td>\n",
       "      <td>595</td>\n",
       "      <td>1062</td>\n",
       "      <td>409</td>\n",
       "      <td>2159</td>\n",
       "      <td>1210</td>\n",
       "      <td>1955</td>\n",
       "      <td>808</td>\n",
       "      <td>1997</td>\n",
       "      <td>711</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25839</th>\n",
       "      <td>CAATGGCCCTGGGACACAGACGTGGAGGCAGAGATTACAGCGAAGC...</td>\n",
       "      <td>chr16_29145942_29145961_mut</td>\n",
       "      <td>chr16_29145942_29145961</td>\n",
       "      <td>mut</td>\n",
       "      <td>913</td>\n",
       "      <td>804</td>\n",
       "      <td>569</td>\n",
       "      <td>1217</td>\n",
       "      <td>1556</td>\n",
       "      <td>649</td>\n",
       "      <td>1819</td>\n",
       "      <td>2182</td>\n",
       "      <td>2012</td>\n",
       "      <td>688</td>\n",
       "      <td>1293</td>\n",
       "      <td>1503</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12884</th>\n",
       "      <td>ACTAGGAGTGATTTAAACTACGTGGACATTTATTATCTCCTGTATT...</td>\n",
       "      <td>chr3_98778884_98778908_WT</td>\n",
       "      <td>chr3_98778884_98778908</td>\n",
       "      <td>WT</td>\n",
       "      <td>1525</td>\n",
       "      <td>1955</td>\n",
       "      <td>1767</td>\n",
       "      <td>3538</td>\n",
       "      <td>1933</td>\n",
       "      <td>988</td>\n",
       "      <td>4389</td>\n",
       "      <td>4406</td>\n",
       "      <td>6156</td>\n",
       "      <td>2316</td>\n",
       "      <td>5640</td>\n",
       "      <td>3157</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41945</th>\n",
       "      <td>TTCCTGGAAGAGGGTGGAAGTGAGATCAGTACAGACACGTGATCTG...</td>\n",
       "      <td>chr20_35318743_35318757_shuffle</td>\n",
       "      <td>chr20_35318743_35318757</td>\n",
       "      <td>shuffle</td>\n",
       "      <td>344</td>\n",
       "      <td>530</td>\n",
       "      <td>560</td>\n",
       "      <td>789</td>\n",
       "      <td>2088</td>\n",
       "      <td>583</td>\n",
       "      <td>827</td>\n",
       "      <td>1662</td>\n",
       "      <td>1755</td>\n",
       "      <td>550</td>\n",
       "      <td>1488</td>\n",
       "      <td>805</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83946</th>\n",
       "      <td>CGTGGAAAGTGCTTTTGGAAAAGTTGCCGCGGGATTGTGGTAGAGC...</td>\n",
       "      <td>chr13_40252431_40252450_fullShuffle</td>\n",
       "      <td>chr13_40252431_40252450</td>\n",
       "      <td>fullShuffle</td>\n",
       "      <td>452</td>\n",
       "      <td>1098</td>\n",
       "      <td>999</td>\n",
       "      <td>1342</td>\n",
       "      <td>1876</td>\n",
       "      <td>839</td>\n",
       "      <td>3229</td>\n",
       "      <td>2769</td>\n",
       "      <td>3222</td>\n",
       "      <td>1232</td>\n",
       "      <td>1327</td>\n",
       "      <td>1104</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                     seq  \\\n",
       "69090  ACGAGTCCGCTGAAGACCGTCACAACTTCCGTCATCCGTCCGCCTA...   \n",
       "25839  CAATGGCCCTGGGACACAGACGTGGAGGCAGAGATTACAGCGAAGC...   \n",
       "12884  ACTAGGAGTGATTTAAACTACGTGGACATTTATTATCTCCTGTATT...   \n",
       "41945  TTCCTGGAAGAGGGTGGAAGTGAGATCAGTACAGACACGTGATCTG...   \n",
       "83946  CGTGGAAAGTGCTTTTGGAAAAGTTGCCGCGGGATTGTGGTAGAGC...   \n",
       "\n",
       "                                        id                      loc  \\\n",
       "69090  chr9_13027983_13028007_flankShuffle   chr9_13027983_13028007   \n",
       "25839          chr16_29145942_29145961_mut  chr16_29145942_29145961   \n",
       "12884            chr3_98778884_98778908_WT   chr3_98778884_98778908   \n",
       "41945      chr20_35318743_35318757_shuffle  chr20_35318743_35318757   \n",
       "83946  chr13_40252431_40252450_fullShuffle  chr13_40252431_40252450   \n",
       "\n",
       "      enhancer_variant  HaCaT_1_count  HaCaT_2_count  MCF10AGus_count  \\\n",
       "69090     flankShuffle            559            958              927   \n",
       "25839              mut            913            804              569   \n",
       "12884               WT           1525           1955             1767   \n",
       "41945          shuffle            344            530              560   \n",
       "83946      fullShuffle            452           1098              999   \n",
       "\n",
       "       MCF10Ap53KO_1_count  MCF10Ap53KO_2_count  MCF10ATAp63B_count  \\\n",
       "69090                  595                 1062                 409   \n",
       "25839                 1217                 1556                 649   \n",
       "12884                 3538                 1933                 988   \n",
       "41945                  789                 2088                 583   \n",
       "83946                 1342                 1876                 839   \n",
       "\n",
       "       MCF10A_1_count  MCF10A_2_count  plasmid_1_count  plasmid_2_count  \\\n",
       "69090            2159            1210             1955              808   \n",
       "25839            1819            2182             2012              688   \n",
       "12884            4389            4406             6156             2316   \n",
       "41945             827            1662             1755              550   \n",
       "83946            3229            2769             3222             1232   \n",
       "\n",
       "       SCC25_1_count  SCC25_2_count  \n",
       "69090           1997            711  \n",
       "25839           1293           1503  \n",
       "12884           5640           3157  \n",
       "41945           1488            805  \n",
       "83946           1327           1104  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#import libraries\n",
    "import pandas as pd\n",
    "import os\n",
    "#import key tables\n",
    "os.chdir(r'Z:\\gbaniulyte\\MPRA\\AM_STARRSeq\\final_scripts_fastq\\tables')\n",
    "counts_path = r'Z:\\gbaniulyte\\MPRA\\AM_STARRSeq\\final_scripts_fastq\\tables\\read_counts'\n",
    "#this is the original table with seq information\n",
    "df = pd.read_csv('p63enh_starrseq_info_long.csv', usecols=['seq', 'id'])\n",
    "df[['loc','enhancer_variant']] = df.apply(lambda row: [row.id[:row.id.rfind('_')],\n",
    "\t\t\t\t\t\t\t\t\t\t\t\t\t   row.id[row.id.rfind('_') + 1 :]],\n",
    "\t\t\t\t\t\t\t\t\t\t  \t\t\t\taxis='columns', result_type='expand')\n",
    "df.seq = df.seq.apply(lambda x: x[:101].upper())\n",
    "#gather all the counts tables for each sample\n",
    "count_tbl = os.listdir(counts_path)\n",
    "for tbl in count_tbl:\n",
    "\tcounts = pd.read_csv(os.path.join(counts_path,tbl), names=['seq',tbl[:tbl.find('.csv')]])\n",
    "\tdf = df.merge(counts, how='left', on='seq')\n",
    "#df.to_csv('p63enh_starrseq_raw_counts.csv')\n",
    "df.sample(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Prepare normalized counts table from all p63 enhancer STARRseq experiments.\n",
    "- Used minimal read threshold as 5 for DNA dn RNA reads\n",
    "- Normalize each sample to total reads in that sample pool\n",
    "- Then normalize RNA reads (MCF10A, MCF10Ap53KO, HaCaT, SCC25, MCF10A + Gus & MCF10A + TAp63β) DNA reads.\n",
    "- All cell types except HaCaT were trasnfected with plasmid pool 1 so they are normalized to plasmid 1, HaCaTs normalized to plasmid 2 DNA counts.\n",
    "- Create separate raw read-filtered tables as we're only considering 2 or 3 cell types for each figures and SCC25 cells trasfected poorly and would lower usable enhacner numbers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of enhancers before filtering: 86550\n",
      "Number of enhancers after filtering for MCF10A WT and p53KO: 81867\n",
      "Number of enhancers after filtering for MCF10A p53KO, HaCaT and SCC25: 58259\n",
      "Number of enhancers after filtering for MCF10A Gus and TAp63B: 83167\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>loc</th>\n",
       "      <th>enhancer_variant</th>\n",
       "      <th>RNA/DNA_MCF10A_1</th>\n",
       "      <th>RNA/DNA_MCF10A_2</th>\n",
       "      <th>RNA/DNA_MCF10Ap53KO_1</th>\n",
       "      <th>RNA/DNA_MCF10Ap53KO_2</th>\n",
       "      <th>RNA/DNA_MCF10AGus</th>\n",
       "      <th>RNA/DNA_MCF10ATAp63B</th>\n",
       "      <th>RNA/DNA_SCC25_1</th>\n",
       "      <th>RNA/DNA_SCC25_2</th>\n",
       "      <th>RNA/DNA_HaCaT_1</th>\n",
       "      <th>RNA/DNA_HaCaT_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>13869</th>\n",
       "      <td>chr1_40890062_40890076_WT</td>\n",
       "      <td>chr1_40890062_40890076</td>\n",
       "      <td>WT</td>\n",
       "      <td>0.740127</td>\n",
       "      <td>0.528800</td>\n",
       "      <td>0.432679</td>\n",
       "      <td>0.475164</td>\n",
       "      <td>0.739241</td>\n",
       "      <td>0.188971</td>\n",
       "      <td>0.489176</td>\n",
       "      <td>0.785243</td>\n",
       "      <td>0.429115</td>\n",
       "      <td>0.510577</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55817</th>\n",
       "      <td>chr6_161375409_161375428_flankShuffle</td>\n",
       "      <td>chr6_161375409_161375428</td>\n",
       "      <td>flankShuffle</td>\n",
       "      <td>0.688981</td>\n",
       "      <td>0.653830</td>\n",
       "      <td>1.107887</td>\n",
       "      <td>0.669541</td>\n",
       "      <td>0.689230</td>\n",
       "      <td>0.947056</td>\n",
       "      <td>0.463220</td>\n",
       "      <td>0.761670</td>\n",
       "      <td>0.696045</td>\n",
       "      <td>0.548090</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83475</th>\n",
       "      <td>chr10_79454556_79454575_fullShuffle</td>\n",
       "      <td>chr10_79454556_79454575</td>\n",
       "      <td>fullShuffle</td>\n",
       "      <td>0.584642</td>\n",
       "      <td>0.657283</td>\n",
       "      <td>0.470897</td>\n",
       "      <td>0.982210</td>\n",
       "      <td>0.869058</td>\n",
       "      <td>0.504326</td>\n",
       "      <td>0.291994</td>\n",
       "      <td>1.044609</td>\n",
       "      <td>1.461568</td>\n",
       "      <td>1.234655</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73705</th>\n",
       "      <td>chr16_53768760_53768784_fullShuffle</td>\n",
       "      <td>chr16_53768760_53768784</td>\n",
       "      <td>fullShuffle</td>\n",
       "      <td>0.635600</td>\n",
       "      <td>0.387273</td>\n",
       "      <td>0.339047</td>\n",
       "      <td>0.729203</td>\n",
       "      <td>0.600748</td>\n",
       "      <td>0.357635</td>\n",
       "      <td>0.000961</td>\n",
       "      <td>1.021531</td>\n",
       "      <td>0.975654</td>\n",
       "      <td>0.323150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41357</th>\n",
       "      <td>chr10_121781335_121781359_shuffle</td>\n",
       "      <td>chr10_121781335_121781359</td>\n",
       "      <td>shuffle</td>\n",
       "      <td>0.692604</td>\n",
       "      <td>0.635498</td>\n",
       "      <td>0.785038</td>\n",
       "      <td>0.920904</td>\n",
       "      <td>0.803168</td>\n",
       "      <td>0.567769</td>\n",
       "      <td>1.798843</td>\n",
       "      <td>0.931615</td>\n",
       "      <td>0.942481</td>\n",
       "      <td>1.045066</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          id                        loc  \\\n",
       "13869              chr1_40890062_40890076_WT     chr1_40890062_40890076   \n",
       "55817  chr6_161375409_161375428_flankShuffle   chr6_161375409_161375428   \n",
       "83475    chr10_79454556_79454575_fullShuffle    chr10_79454556_79454575   \n",
       "73705    chr16_53768760_53768784_fullShuffle    chr16_53768760_53768784   \n",
       "41357      chr10_121781335_121781359_shuffle  chr10_121781335_121781359   \n",
       "\n",
       "      enhancer_variant  RNA/DNA_MCF10A_1  RNA/DNA_MCF10A_2  \\\n",
       "13869               WT          0.740127          0.528800   \n",
       "55817     flankShuffle          0.688981          0.653830   \n",
       "83475      fullShuffle          0.584642          0.657283   \n",
       "73705      fullShuffle          0.635600          0.387273   \n",
       "41357          shuffle          0.692604          0.635498   \n",
       "\n",
       "       RNA/DNA_MCF10Ap53KO_1  RNA/DNA_MCF10Ap53KO_2  RNA/DNA_MCF10AGus  \\\n",
       "13869               0.432679               0.475164           0.739241   \n",
       "55817               1.107887               0.669541           0.689230   \n",
       "83475               0.470897               0.982210           0.869058   \n",
       "73705               0.339047               0.729203           0.600748   \n",
       "41357               0.785038               0.920904           0.803168   \n",
       "\n",
       "       RNA/DNA_MCF10ATAp63B  RNA/DNA_SCC25_1  RNA/DNA_SCC25_2  \\\n",
       "13869              0.188971         0.489176         0.785243   \n",
       "55817              0.947056         0.463220         0.761670   \n",
       "83475              0.504326         0.291994         1.044609   \n",
       "73705              0.357635         0.000961         1.021531   \n",
       "41357              0.567769         1.798843         0.931615   \n",
       "\n",
       "       RNA/DNA_HaCaT_1  RNA/DNA_HaCaT_2  \n",
       "13869         0.429115         0.510577  \n",
       "55817         0.696045         0.548090  \n",
       "83475         1.461568         1.234655  \n",
       "73705         0.975654         0.323150  \n",
       "41357         0.942481         1.045066  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#import libraries\n",
    "import pandas as pd\n",
    "import os\n",
    "import numpy as np\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "#import key tables\n",
    "os.chdir(r'Z:\\gbaniulyte\\MPRA\\AM_STARRSeq\\final_scripts_fastq\\tables')\n",
    "counts_path = r'Z:\\gbaniulyte\\MPRA\\AM_STARRSeq\\final_scripts_fastq\\tables\\read_counts'\n",
    "min_reads = 5\n",
    "#this is the original table with seq information\n",
    "df = pd.read_csv('p63enh_starrseq_raw_counts.csv', index_col=0)\n",
    "df.drop(columns=['seq'], inplace=True)\n",
    "#get total read counts per pool\n",
    "total_counts = {'MCF10A_1_count':df.MCF10A_1_count.sum(), 'MCF10A_2_count':df.MCF10A_2_count.sum(),\n",
    "\t\t 'MCF10Ap53KO_1_count':df.MCF10Ap53KO_1_count.sum(),'MCF10Ap53KO_2_count':df.MCF10Ap53KO_2_count.sum(),\n",
    "\t\t 'HaCaT_1_count':df.HaCaT_1_count.sum(),'HaCaT_2_count':df.HaCaT_2_count.sum(),\n",
    "\t\t 'SCC25_1_count':df.SCC25_1_count.sum(),'SCC25_2_count':df.SCC25_2_count.sum(),\n",
    "\t\t 'plasmid_1_count':df.plasmid_1_count.sum(),'plasmid_2_count':df.plasmid_2_count.sum(),\n",
    "\t\t 'MCF10AGus_count':df.MCF10AGus_count.sum(), 'MCF10ATAp63B_count':df.MCF10ATAp63B_count.sum()}\n",
    "#organize and name all the columns\n",
    "col_list = {'new_col':['RNA/DNA_MCF10A_1','RNA/DNA_MCF10A_2','RNA/DNA_MCF10Ap53KO_1','RNA/DNA_MCF10Ap53KO_2','RNA/DNA_MCF10AGus','RNA/DNA_MCF10ATAp63B',\n",
    "\t\t\t\t\t    'RNA/DNA_SCC25_1', 'RNA/DNA_SCC25_2','RNA/DNA_HaCaT_1','RNA/DNA_HaCaT_2','Plasmid_1','Plasmid_2'],\n",
    "\t\t\t'count_col':['MCF10A_1_count','MCF10A_2_count','MCF10Ap53KO_1_count','MCF10Ap53KO_2_count','MCF10AGus_count','MCF10ATAp63B_count',\n",
    "\t\t\t\t \t\t 'SCC25_1_count','SCC25_2_count','HaCaT_1_count','HaCaT_2_count','plasmid_1_count','plasmid_2_count']}\n",
    "df_norm = df.copy()\n",
    "#normalize to total reads\n",
    "for n in range(0,len(col_list['new_col'])):\n",
    "\tdf_norm[col_list['new_col'][n]] = df_norm[col_list['count_col'][n]].div(total_counts[col_list['count_col'][n]])\n",
    "#normalize all MCFs to plasmid pool 1\n",
    "df_norm[col_list['new_col'][0:8]] = df_norm[col_list['new_col'][0:8]].div(df_norm['Plasmid_1'].values, axis='index')\n",
    "#fresh pool of DNA was maxipreped for transfection into HaCats\n",
    "df_norm[['RNA/DNA_HaCaT_1','RNA/DNA_HaCaT_2']] = df_norm[['RNA/DNA_HaCaT_1','RNA/DNA_HaCaT_2']].div(\n",
    "\tdf_norm['Plasmid_2'].values, axis='index')\n",
    "#filter out based on min read threshold\n",
    "print('Number of enhancers before filtering:', len(df_norm))\n",
    "#create different tables for matched enhancers because SCC25 had signf. lower read counts\n",
    "df_norm_1 = df_norm[(df_norm.plasmid_1_count > min_reads) &\n",
    "\t\t(df_norm.MCF10A_1_count > min_reads) & (df_norm.MCF10A_2_count > min_reads) &\n",
    " \t\t(df_norm.MCF10Ap53KO_1_count > min_reads) & (df_norm.MCF10Ap53KO_2_count > min_reads)]\n",
    "print('Number of enhancers after filtering for MCF10A WT and p53KO:', len(df_norm_1))\n",
    "df_norm_1.drop(columns=col_list['count_col'] + ['Plasmid_1','Plasmid_2'], inplace=True)\n",
    "df_norm_2 = df_norm[(df_norm.plasmid_1_count > min_reads) & (df_norm.plasmid_2_count > min_reads) &\n",
    " \t\t(df_norm.MCF10Ap53KO_1_count > min_reads) & (df_norm.MCF10Ap53KO_2_count > min_reads) &\n",
    "\t\t(df_norm.HaCaT_1_count > min_reads) & (df_norm.HaCaT_2_count > min_reads) &\n",
    "\t\t(df_norm.SCC25_1_count > min_reads) & (df_norm.SCC25_2_count > min_reads)]\n",
    "print('Number of enhancers after filtering for MCF10A p53KO, HaCaT and SCC25:', len(df_norm_2))\n",
    "df_norm_2.drop(columns=col_list['count_col'] + ['Plasmid_1','Plasmid_2'], inplace=True)\n",
    "df_norm_3 = df_norm[(df_norm.plasmid_1_count > min_reads) &\n",
    "\t\t(df_norm.MCF10AGus_count > min_reads) & (df_norm.MCF10ATAp63B_count > min_reads)]\n",
    "print('Number of enhancers after filtering for MCF10A Gus and TAp63B:', len(df_norm_3))\n",
    "df_norm_3.drop(columns=col_list['count_col'] + ['Plasmid_1','Plasmid_2'], inplace=True)\n",
    "df_norm_3.sample(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Match enhancers for:\n",
    "- Replicates\n",
    "- MCF10A WT and KO for all 5 variants\n",
    "- Separately MCF10A p53KO, HaCaT and SCC25 for WT and mut only\n",
    "- Separately MCF10A Gus or TAp6B WT and Mut"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of unique enhancer locations in MCF10As WT and p53KO: 13696\n",
      "Number of unique enhancer locations in MCF10As p53KO, HaCaT and SCC25: 8566\n",
      "Number of unique enhancer locations in MCF10As Gus and TAp63B: 16143\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>loc</th>\n",
       "      <th colspan=\"2\" halign=\"left\">RNA/DNA_MCF10Ap53KO_1</th>\n",
       "      <th colspan=\"2\" halign=\"left\">RNA/DNA_MCF10Ap53KO_2</th>\n",
       "      <th colspan=\"2\" halign=\"left\">RNA/DNA_HaCaT_1</th>\n",
       "      <th colspan=\"2\" halign=\"left\">RNA/DNA_HaCaT_2</th>\n",
       "      <th colspan=\"2\" halign=\"left\">RNA/DNA_SCC25_1</th>\n",
       "      <th colspan=\"2\" halign=\"left\">RNA/DNA_SCC25_2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>enhancer_variant</th>\n",
       "      <th></th>\n",
       "      <th>WT</th>\n",
       "      <th>mut</th>\n",
       "      <th>WT</th>\n",
       "      <th>mut</th>\n",
       "      <th>WT</th>\n",
       "      <th>mut</th>\n",
       "      <th>WT</th>\n",
       "      <th>mut</th>\n",
       "      <th>WT</th>\n",
       "      <th>mut</th>\n",
       "      <th>WT</th>\n",
       "      <th>mut</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10_100043539_100043558</td>\n",
       "      <td>1.319655</td>\n",
       "      <td>0.560156</td>\n",
       "      <td>1.137255</td>\n",
       "      <td>0.733605</td>\n",
       "      <td>1.058830</td>\n",
       "      <td>0.638685</td>\n",
       "      <td>1.254867</td>\n",
       "      <td>0.879382</td>\n",
       "      <td>0.390366</td>\n",
       "      <td>1.393339</td>\n",
       "      <td>1.058964</td>\n",
       "      <td>0.403181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10_100157341_100157365</td>\n",
       "      <td>1.671221</td>\n",
       "      <td>1.560248</td>\n",
       "      <td>1.620464</td>\n",
       "      <td>1.152133</td>\n",
       "      <td>1.614002</td>\n",
       "      <td>0.918617</td>\n",
       "      <td>1.724374</td>\n",
       "      <td>1.219318</td>\n",
       "      <td>0.596935</td>\n",
       "      <td>0.666459</td>\n",
       "      <td>0.402588</td>\n",
       "      <td>1.351475</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10_100286133_100286152</td>\n",
       "      <td>0.559138</td>\n",
       "      <td>0.383291</td>\n",
       "      <td>0.543610</td>\n",
       "      <td>0.348498</td>\n",
       "      <td>0.416229</td>\n",
       "      <td>0.338045</td>\n",
       "      <td>0.484690</td>\n",
       "      <td>0.366422</td>\n",
       "      <td>1.037235</td>\n",
       "      <td>0.516950</td>\n",
       "      <td>0.841603</td>\n",
       "      <td>0.651545</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr10_100373316_100373330</td>\n",
       "      <td>3.184686</td>\n",
       "      <td>1.870014</td>\n",
       "      <td>2.757653</td>\n",
       "      <td>1.644806</td>\n",
       "      <td>1.168821</td>\n",
       "      <td>1.835282</td>\n",
       "      <td>1.100346</td>\n",
       "      <td>1.115179</td>\n",
       "      <td>1.876799</td>\n",
       "      <td>1.543624</td>\n",
       "      <td>2.005602</td>\n",
       "      <td>3.341386</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr10_100886190_100886214</td>\n",
       "      <td>0.464043</td>\n",
       "      <td>0.313704</td>\n",
       "      <td>0.345795</td>\n",
       "      <td>0.545380</td>\n",
       "      <td>0.827458</td>\n",
       "      <td>1.056786</td>\n",
       "      <td>0.841565</td>\n",
       "      <td>1.224738</td>\n",
       "      <td>1.814806</td>\n",
       "      <td>1.167748</td>\n",
       "      <td>0.511488</td>\n",
       "      <td>0.283157</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        loc RNA/DNA_MCF10Ap53KO_1            \\\n",
       "enhancer_variant                                               WT       mut   \n",
       "0                 chr10_100043539_100043558              1.319655  0.560156   \n",
       "1                 chr10_100157341_100157365              1.671221  1.560248   \n",
       "2                 chr10_100286133_100286152              0.559138  0.383291   \n",
       "3                 chr10_100373316_100373330              3.184686  1.870014   \n",
       "6                 chr10_100886190_100886214              0.464043  0.313704   \n",
       "\n",
       "                 RNA/DNA_MCF10Ap53KO_2           RNA/DNA_HaCaT_1            \\\n",
       "enhancer_variant                    WT       mut              WT       mut   \n",
       "0                             1.137255  0.733605        1.058830  0.638685   \n",
       "1                             1.620464  1.152133        1.614002  0.918617   \n",
       "2                             0.543610  0.348498        0.416229  0.338045   \n",
       "3                             2.757653  1.644806        1.168821  1.835282   \n",
       "6                             0.345795  0.545380        0.827458  1.056786   \n",
       "\n",
       "                 RNA/DNA_HaCaT_2           RNA/DNA_SCC25_1            \\\n",
       "enhancer_variant              WT       mut              WT       mut   \n",
       "0                       1.254867  0.879382        0.390366  1.393339   \n",
       "1                       1.724374  1.219318        0.596935  0.666459   \n",
       "2                       0.484690  0.366422        1.037235  0.516950   \n",
       "3                       1.100346  1.115179        1.876799  1.543624   \n",
       "6                       0.841565  1.224738        1.814806  1.167748   \n",
       "\n",
       "                 RNA/DNA_SCC25_2            \n",
       "enhancer_variant              WT       mut  \n",
       "0                       1.058964  0.403181  \n",
       "1                       0.402588  1.351475  \n",
       "2                       0.841603  0.651545  \n",
       "3                       2.005602  3.341386  \n",
       "6                       0.511488  0.283157  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#move enh variant into columns to make it easier to match enahncers in each sample pool\n",
    "#MCF10A WT and KO, all variants\n",
    "wide_mcf = df_norm_1.pivot(index=['loc'], columns=['enhancer_variant'],\n",
    "\t\t values=['RNA/DNA_MCF10A_1','RNA/DNA_MCF10A_2',\n",
    "\t\t   'RNA/DNA_MCF10Ap53KO_1','RNA/DNA_MCF10Ap53KO_2']).reset_index()\n",
    "#MCF10A p53KO, HaCaT, SCC25, WT and mut variants\n",
    "wide_mhs = df_norm_2[df_norm_2.enhancer_variant.isin(['WT','mut'])].pivot(index=['loc'], columns=['enhancer_variant'],\n",
    "\t\t values=['RNA/DNA_MCF10Ap53KO_1','RNA/DNA_MCF10Ap53KO_2',\n",
    "\t\t   'RNA/DNA_HaCaT_1','RNA/DNA_HaCaT_2',\n",
    "\t\t   'RNA/DNA_SCC25_1','RNA/DNA_SCC25_2']).reset_index()\n",
    "#MCF10A Gus and TAp63B overexpression, WT and mut variants\n",
    "wide_overex = df_norm_3[df_norm_3.enhancer_variant.isin(['WT','mut'])].pivot(index=['loc'], columns=['enhancer_variant'],\n",
    "\t\t values=['RNA/DNA_MCF10AGus','RNA/DNA_MCF10ATAp63B']).reset_index()\n",
    "#drop NaNs in each row to remove enhancers that don't have a match in replicate/cell type/enh variant\n",
    "wide_mcf.dropna(how='any', inplace=True)\n",
    "wide_mhs.dropna(how='any', inplace=True)\n",
    "wide_overex.dropna(how='any', inplace=True)\n",
    "print('Number of unique enhancer locations in MCF10As WT and p53KO:', wide_mcf['loc',''].nunique())\n",
    "print('Number of unique enhancer locations in MCF10As p53KO, HaCaT and SCC25:', wide_mhs['loc',''].nunique())\n",
    "print('Number of unique enhancer locations in MCF10As Gus and TAp63B:', wide_overex['loc',''].nunique())\n",
    "wide_mhs.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Average all the replicates and create a long table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>loc</th>\n",
       "      <th>enhancer_variant</th>\n",
       "      <th>RNA/DNA_MCF10AGus</th>\n",
       "      <th>RNA/DNA_MCF10ATAp63B</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1_3717092_3717116_WT</td>\n",
       "      <td>chr1_3717092_3717116</td>\n",
       "      <td>WT</td>\n",
       "      <td>0.565690</td>\n",
       "      <td>1.798371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1_5652400_5652424_WT</td>\n",
       "      <td>chr1_5652400_5652424</td>\n",
       "      <td>WT</td>\n",
       "      <td>1.387837</td>\n",
       "      <td>4.440618</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1_23167597_23167616_WT</td>\n",
       "      <td>chr1_23167597_23167616</td>\n",
       "      <td>WT</td>\n",
       "      <td>1.578312</td>\n",
       "      <td>4.488055</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1_31575754_31575773_WT</td>\n",
       "      <td>chr1_31575754_31575773</td>\n",
       "      <td>WT</td>\n",
       "      <td>1.236709</td>\n",
       "      <td>0.889068</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1_38027284_38027308_WT</td>\n",
       "      <td>chr1_38027284_38027308</td>\n",
       "      <td>WT</td>\n",
       "      <td>1.133709</td>\n",
       "      <td>3.157620</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34615</th>\n",
       "      <td>chr9_135951083_135951107_mut</td>\n",
       "      <td>chr9_135951083_135951107</td>\n",
       "      <td>mut</td>\n",
       "      <td>0.067028</td>\n",
       "      <td>0.049090</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34616</th>\n",
       "      <td>chr9_136484776_136484795_mut</td>\n",
       "      <td>chr9_136484776_136484795</td>\n",
       "      <td>mut</td>\n",
       "      <td>0.523268</td>\n",
       "      <td>0.473875</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34617</th>\n",
       "      <td>chr9_136533879_136533893_mut</td>\n",
       "      <td>chr9_136533879_136533893</td>\n",
       "      <td>mut</td>\n",
       "      <td>0.896743</td>\n",
       "      <td>1.007147</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34618</th>\n",
       "      <td>chr9_137062571_137062595_mut</td>\n",
       "      <td>chr9_137062571_137062595</td>\n",
       "      <td>mut</td>\n",
       "      <td>1.814900</td>\n",
       "      <td>1.955652</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34619</th>\n",
       "      <td>chr9_137310751_137310775_mut</td>\n",
       "      <td>chr9_137310751_137310775</td>\n",
       "      <td>mut</td>\n",
       "      <td>0.557032</td>\n",
       "      <td>0.702577</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>32286 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 id                       loc  \\\n",
       "0           chr1_3717092_3717116_WT      chr1_3717092_3717116   \n",
       "1           chr1_5652400_5652424_WT      chr1_5652400_5652424   \n",
       "2         chr1_23167597_23167616_WT    chr1_23167597_23167616   \n",
       "3         chr1_31575754_31575773_WT    chr1_31575754_31575773   \n",
       "4         chr1_38027284_38027308_WT    chr1_38027284_38027308   \n",
       "...                             ...                       ...   \n",
       "34615  chr9_135951083_135951107_mut  chr9_135951083_135951107   \n",
       "34616  chr9_136484776_136484795_mut  chr9_136484776_136484795   \n",
       "34617  chr9_136533879_136533893_mut  chr9_136533879_136533893   \n",
       "34618  chr9_137062571_137062595_mut  chr9_137062571_137062595   \n",
       "34619  chr9_137310751_137310775_mut  chr9_137310751_137310775   \n",
       "\n",
       "      enhancer_variant  RNA/DNA_MCF10AGus  RNA/DNA_MCF10ATAp63B  \n",
       "0                   WT           0.565690              1.798371  \n",
       "1                   WT           1.387837              4.440618  \n",
       "2                   WT           1.578312              4.488055  \n",
       "3                   WT           1.236709              0.889068  \n",
       "4                   WT           1.133709              3.157620  \n",
       "...                ...                ...                   ...  \n",
       "34615              mut           0.067028              0.049090  \n",
       "34616              mut           0.523268              0.473875  \n",
       "34617              mut           0.896743              1.007147  \n",
       "34618              mut           1.814900              1.955652  \n",
       "34619              mut           0.557032              0.702577  \n",
       "\n",
       "[32286 rows x 5 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#use the wide tables to get a list of enahcners to keep (matched enhancers)\n",
    "to_keep_mcf = wide_mcf['loc', ''].tolist()\n",
    "to_keep_mhs = wide_mhs['loc', ''].tolist()\n",
    "to_keep_overex = wide_overex['loc', ''].tolist()\n",
    "#average replicates for each cell type\n",
    "df_norm_2 = df_norm_2[df_norm_2.enhancer_variant.isin(['WT', 'mut'])]\n",
    "df_norm_3 = df_norm_3[df_norm_3.enhancer_variant.isin(['WT', 'mut'])]\n",
    "df_norm_1 = df_norm_1[df_norm_1['loc'].isin(to_keep_mcf)]\n",
    "df_norm_2 = df_norm_2[df_norm_2['loc'].isin(to_keep_mhs)]\n",
    "df_norm_3 = df_norm_3[df_norm_3['loc'].isin(to_keep_overex)]\n",
    "df_norm_1['RNA/DNA_MCF10A'] = df_norm_1[['RNA/DNA_MCF10A_1','RNA/DNA_MCF10A_2']].mean(axis=1)\n",
    "df_norm_1['RNA/DNA_MCF10Ap53KO'] = df_norm_1[['RNA/DNA_MCF10Ap53KO_1','RNA/DNA_MCF10Ap53KO_2']].mean(axis=1)\n",
    "df_norm_2['RNA/DNA_MCF10Ap53KO'] = df_norm_2[['RNA/DNA_MCF10Ap53KO_1','RNA/DNA_MCF10Ap53KO_2']].mean(axis=1)\n",
    "df_norm_2['RNA/DNA_HaCaT'] = df_norm_2[['RNA/DNA_HaCaT_1','RNA/DNA_HaCaT_2']].mean(axis=1)\n",
    "df_norm_2['RNA/DNA_SCC25'] = df_norm_2[['RNA/DNA_SCC25_1','RNA/DNA_SCC25_2']].mean(axis=1)\n",
    "#Gus and TAp63B overexpression is only a single replicate\n",
    "df_norm_1.drop(columns=col_list['new_col'][0:-2],inplace=True)\n",
    "df_norm_2.drop(columns=col_list['new_col'][0:-2],inplace=True)\n",
    "df_norm_3.drop(columns=['RNA/DNA_MCF10A_1','RNA/DNA_MCF10A_2','RNA/DNA_MCF10Ap53KO_1','RNA/DNA_MCF10Ap53KO_2',\n",
    "\t\t\t\t\t    'RNA/DNA_SCC25_1', 'RNA/DNA_SCC25_2','RNA/DNA_HaCaT_1','RNA/DNA_HaCaT_2'],inplace=True)\n",
    "df_norm_3"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Finalize tables\n",
    "- Add log2(WT/mut) scores to measure p63RE-dependent enhancer activity\n",
    "- Add categorical information: motif type, class, observation scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>loc</th>\n",
       "      <th>enhancer_variant</th>\n",
       "      <th>cell_line</th>\n",
       "      <th>RNA/DNA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1_3717092_3717116</td>\n",
       "      <td>WT</td>\n",
       "      <td>MCF10A</td>\n",
       "      <td>0.704075</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1_31575754_31575773</td>\n",
       "      <td>WT</td>\n",
       "      <td>MCF10A</td>\n",
       "      <td>1.586092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1_38117813_38117832</td>\n",
       "      <td>WT</td>\n",
       "      <td>MCF10A</td>\n",
       "      <td>1.681971</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1_40886808_40886827</td>\n",
       "      <td>WT</td>\n",
       "      <td>MCF10A</td>\n",
       "      <td>2.157630</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1_42941183_42941207</td>\n",
       "      <td>WT</td>\n",
       "      <td>MCF10A</td>\n",
       "      <td>1.995754</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136955</th>\n",
       "      <td>chr9_132993510_132993534</td>\n",
       "      <td>fullShuffle</td>\n",
       "      <td>MCF10A p53KO</td>\n",
       "      <td>0.674189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136956</th>\n",
       "      <td>chr9_135016437_135016456</td>\n",
       "      <td>fullShuffle</td>\n",
       "      <td>MCF10A p53KO</td>\n",
       "      <td>1.347042</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136957</th>\n",
       "      <td>chr9_136533879_136533893</td>\n",
       "      <td>fullShuffle</td>\n",
       "      <td>MCF10A p53KO</td>\n",
       "      <td>1.351870</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136958</th>\n",
       "      <td>chr9_137062571_137062595</td>\n",
       "      <td>fullShuffle</td>\n",
       "      <td>MCF10A p53KO</td>\n",
       "      <td>1.612355</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136959</th>\n",
       "      <td>chr9_137310751_137310775</td>\n",
       "      <td>fullShuffle</td>\n",
       "      <td>MCF10A p53KO</td>\n",
       "      <td>1.008244</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>136960 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                             loc enhancer_variant     cell_line   RNA/DNA\n",
       "0           chr1_3717092_3717116               WT        MCF10A  0.704075\n",
       "1         chr1_31575754_31575773               WT        MCF10A  1.586092\n",
       "2         chr1_38117813_38117832               WT        MCF10A  1.681971\n",
       "3         chr1_40886808_40886827               WT        MCF10A  2.157630\n",
       "4         chr1_42941183_42941207               WT        MCF10A  1.995754\n",
       "...                          ...              ...           ...       ...\n",
       "136955  chr9_132993510_132993534      fullShuffle  MCF10A p53KO  0.674189\n",
       "136956  chr9_135016437_135016456      fullShuffle  MCF10A p53KO  1.347042\n",
       "136957  chr9_136533879_136533893      fullShuffle  MCF10A p53KO  1.351870\n",
       "136958  chr9_137062571_137062595      fullShuffle  MCF10A p53KO  1.612355\n",
       "136959  chr9_137310751_137310775      fullShuffle  MCF10A p53KO  1.008244\n",
       "\n",
       "[136960 rows x 4 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#collect cell type information into the column\n",
    "df_norm_long_1 = df_norm_1.melt(id_vars=['loc','enhancer_variant'],\n",
    "\t\t\t\t\t\t\t\t\tvalue_vars=['RNA/DNA_MCF10A','RNA/DNA_MCF10Ap53KO'],\n",
    "\t\t\t\t\t\t\t\t\tvalue_name='RNA/DNA', var_name='cell_line')\n",
    "df_norm_long_2 = df_norm_2.melt(id_vars=['loc','enhancer_variant'],\n",
    "\t\t\t\t\t\t\t\t\tvalue_vars=['RNA/DNA_MCF10Ap53KO','RNA/DNA_HaCaT','RNA/DNA_SCC25'],\n",
    "\t\t\t\t\t\t\t\t\tvalue_name='RNA/DNA', var_name='cell_line')\n",
    "df_norm_long_3 = df_norm_3.melt(id_vars=['loc','enhancer_variant'],\n",
    "\t\t\t\t\t\t\t\t\tvalue_vars=['RNA/DNA_MCF10AGus','RNA/DNA_MCF10ATAp63B'],\n",
    "\t\t\t\t\t\t\t\t\tvalue_name='RNA/DNA', var_name='cell_line')\n",
    "replace_dict = {'RNA/DNA_MCF10A':'MCF10A', 'RNA/DNA_MCF10Ap53KO':'MCF10A p53KO','RNA/DNA_MCF10AGus':'Gus','RNA/DNA_MCF10ATAp63B':'TAp63B',\n",
    "\t\t\t\t\t\t 'RNA/DNA_HaCaT':'HaCaT','RNA/DNA_SCC25':'SCC25'}\n",
    "df_norm_long_1.replace(replace_dict, inplace=True)\n",
    "df_norm_long_2.replace(replace_dict, inplace=True)\n",
    "df_norm_long_3.replace(replace_dict, inplace=True)\n",
    "df_norm_long_1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>loc</th>\n",
       "      <th>cell_line</th>\n",
       "      <th>WT</th>\n",
       "      <th>mut</th>\n",
       "      <th>log2(WT/mut)</th>\n",
       "      <th>activity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10_100043539_100043558</td>\n",
       "      <td>HaCaT</td>\n",
       "      <td>1.156849</td>\n",
       "      <td>0.759034</td>\n",
       "      <td>0.607964</td>\n",
       "      <td>Activating</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10_100043539_100043558</td>\n",
       "      <td>MCF10A p53KO</td>\n",
       "      <td>1.228455</td>\n",
       "      <td>0.646880</td>\n",
       "      <td>0.925274</td>\n",
       "      <td>Activating</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10_100043539_100043558</td>\n",
       "      <td>SCC25</td>\n",
       "      <td>0.724665</td>\n",
       "      <td>0.898260</td>\n",
       "      <td>-0.309819</td>\n",
       "      <td>Unchanged</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr10_100157341_100157365</td>\n",
       "      <td>HaCaT</td>\n",
       "      <td>1.669188</td>\n",
       "      <td>1.068968</td>\n",
       "      <td>0.642928</td>\n",
       "      <td>Activating</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10_100157341_100157365</td>\n",
       "      <td>MCF10A p53KO</td>\n",
       "      <td>1.645843</td>\n",
       "      <td>1.356191</td>\n",
       "      <td>0.279266</td>\n",
       "      <td>Unchanged</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25693</th>\n",
       "      <td>chr9_99129212_99129236</td>\n",
       "      <td>MCF10A p53KO</td>\n",
       "      <td>0.773522</td>\n",
       "      <td>0.691637</td>\n",
       "      <td>0.161428</td>\n",
       "      <td>Unchanged</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25694</th>\n",
       "      <td>chr9_99129212_99129236</td>\n",
       "      <td>SCC25</td>\n",
       "      <td>1.349398</td>\n",
       "      <td>0.621413</td>\n",
       "      <td>1.118691</td>\n",
       "      <td>Activating</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25695</th>\n",
       "      <td>chr9_99445438_99445462</td>\n",
       "      <td>HaCaT</td>\n",
       "      <td>1.401739</td>\n",
       "      <td>1.055114</td>\n",
       "      <td>0.409819</td>\n",
       "      <td>Unchanged</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25696</th>\n",
       "      <td>chr9_99445438_99445462</td>\n",
       "      <td>MCF10A p53KO</td>\n",
       "      <td>0.746343</td>\n",
       "      <td>0.534718</td>\n",
       "      <td>0.481060</td>\n",
       "      <td>Unchanged</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25697</th>\n",
       "      <td>chr9_99445438_99445462</td>\n",
       "      <td>SCC25</td>\n",
       "      <td>0.716886</td>\n",
       "      <td>0.695675</td>\n",
       "      <td>0.043330</td>\n",
       "      <td>Unchanged</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>25698 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                             loc     cell_line        WT       mut  \\\n",
       "0      chr10_100043539_100043558         HaCaT  1.156849  0.759034   \n",
       "1      chr10_100043539_100043558  MCF10A p53KO  1.228455  0.646880   \n",
       "2      chr10_100043539_100043558         SCC25  0.724665  0.898260   \n",
       "3      chr10_100157341_100157365         HaCaT  1.669188  1.068968   \n",
       "4      chr10_100157341_100157365  MCF10A p53KO  1.645843  1.356191   \n",
       "...                          ...           ...       ...       ...   \n",
       "25693     chr9_99129212_99129236  MCF10A p53KO  0.773522  0.691637   \n",
       "25694     chr9_99129212_99129236         SCC25  1.349398  0.621413   \n",
       "25695     chr9_99445438_99445462         HaCaT  1.401739  1.055114   \n",
       "25696     chr9_99445438_99445462  MCF10A p53KO  0.746343  0.534718   \n",
       "25697     chr9_99445438_99445462         SCC25  0.716886  0.695675   \n",
       "\n",
       "       log2(WT/mut)    activity  \n",
       "0          0.607964  Activating  \n",
       "1          0.925274  Activating  \n",
       "2         -0.309819   Unchanged  \n",
       "3          0.642928  Activating  \n",
       "4          0.279266   Unchanged  \n",
       "...             ...         ...  \n",
       "25693      0.161428   Unchanged  \n",
       "25694      1.118691  Activating  \n",
       "25695      0.409819   Unchanged  \n",
       "25696      0.481060   Unchanged  \n",
       "25697      0.043330   Unchanged  \n",
       "\n",
       "[25698 rows x 6 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#collect enhancer variant information so it's easier to get WT/mut ratio\n",
    "df_wide1 = df_norm_long_1[df_norm_long_1.enhancer_variant.isin(['WT','mut'])].pivot(index=['loc', 'cell_line'],\n",
    "\t\t\t\t\t\t\t\tcolumns=['enhancer_variant'], values=['RNA/DNA']).reset_index()\n",
    "df_wide2 = df_norm_long_2[df_norm_long_2.enhancer_variant.isin(['WT','mut'])].pivot(index=['loc', 'cell_line'],\n",
    "\t\t\t\t\t\t\t\tcolumns=['enhancer_variant'], values=['RNA/DNA']).reset_index()\n",
    "df_wide3 = df_norm_long_3[df_norm_long_3.enhancer_variant.isin(['WT','mut'])].pivot(index=['loc', 'cell_line'],\n",
    "\t\t\t\t\t\t\t\tcolumns=['enhancer_variant'], values=['RNA/DNA']).reset_index()\n",
    "df_wide1['log2(WT/mut)'] = np.log2(df_wide1[('RNA/DNA', 'WT')] / df_wide1[('RNA/DNA', 'mut')])\n",
    "df_wide2['log2(WT/mut)'] = np.log2(df_wide2[('RNA/DNA', 'WT')] / df_wide2[('RNA/DNA', 'mut')])\n",
    "df_wide3['log2(WT/mut)'] = np.log2(df_wide3[('RNA/DNA', 'WT')] / df_wide3[('RNA/DNA', 'mut')])\n",
    "#add categorical labels for FC cutoffs\n",
    "fc_cutoff = 0.58 # == 1.5 FC\n",
    "def dir(ratio):\n",
    "\tif ratio >= fc_cutoff: out = 'Activating'\n",
    "\telif ratio <= -1*fc_cutoff: out = 'Repressing'\n",
    "\telse: out = 'Unchanged'\n",
    "\treturn out\n",
    "for tbl in [df_wide1, df_wide2, df_wide3]:\n",
    "\ttbl.columns = tbl.columns.droplevel(level=1)\n",
    "\ttbl.columns = ['loc', 'cell_line', 'WT', 'mut', 'log2(WT/mut)']\n",
    "\ttbl['activity'] = tbl['log2(WT/mut)'].apply(lambda x: dir(x))\n",
    "#add activity column (log2(WT/mut)) to the normalized expression table\n",
    "df_norm_long_1 = df_norm_long_1.merge(df_wide1[['loc','cell_line','activity']], how='left', on=['loc', 'cell_line'])\n",
    "df_norm_long_2 = df_norm_long_2.merge(df_wide2[['loc','cell_line','activity']], how='left', on=['loc', 'cell_line'])\n",
    "df_norm_long_3 = df_norm_long_3.merge(df_wide3[['loc','cell_line','activity']], how='left', on=['loc', 'cell_line'])\n",
    "df_wide2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "#add categorical info about the enhancers from the table that was cleaned up earlier\n",
    "df_info = pd.read_csv('p63enh_starrseq_info_long.csv', usecols=['obs_score','p63RE_class','strand','loc', 'enhancer_variant','p63RE_type','gc','Number of supporting p53 ChIP-seq datasets_p53'])\n",
    "df_norm_long_1 = df_norm_long_1.merge(df_info, how='left', on=['loc', 'enhancer_variant'])\n",
    "df_norm_long_2 = df_norm_long_2.merge(df_info, how='left', on=['loc', 'enhancer_variant'])\n",
    "df_norm_long_3 = df_norm_long_3.merge(df_info, how='left', on=['loc', 'enhancer_variant'])\n",
    "df_norm_long_1.rename(columns={'Number of supporting p53 ChIP-seq datasets_p53':'obs_p53'}, inplace=True)\n",
    "df_norm_long_2.rename(columns={'Number of supporting p53 ChIP-seq datasets_p53':'obs_p53'}, inplace=True)\n",
    "df_norm_long_3.rename(columns={'Number of supporting p53 ChIP-seq datasets_p53':'obs_p53'}, inplace=True)\n",
    "# #remove irrelevant columns for merge onto fold-change table\n",
    "df_info = df_info[df_info.enhancer_variant == 'WT']\n",
    "df_info.drop(columns=['enhancer_variant','gc'], inplace=True)\n",
    "df_wide1 = df_wide1.merge(df_info, how='left', on='loc')\n",
    "df_wide2 = df_wide2.merge(df_info, how='left', on='loc')\n",
    "df_wide3 = df_wide3.merge(df_info, how='left', on='loc')\n",
    "df_wide1.rename(columns={'Number of supporting p53 ChIP-seq datasets_p53':'obs_p53'}, inplace=True)\n",
    "df_wide2.rename(columns={'Number of supporting p53 ChIP-seq datasets_p53':'obs_p53'}, inplace=True)\n",
    "df_wide3.rename(columns={'Number of supporting p53 ChIP-seq datasets_p53':'obs_p53'}, inplace=True)\n",
    "df_norm_long_1.to_csv('p63enh_starrseq_normCounts_matched_MCF10A_WTp53KO_5variants.csv')\n",
    "df_norm_long_2.to_csv('p63enh_starrseq_normCounts_matched_MCF10Ap53KO_HaCaT_SCC25_WTmut.csv')\n",
    "df_norm_long_3.to_csv('p63enh_starrseq_normCounts_matched_MCF10AGus_TAp63B_WTmut.csv')\n",
    "df_wide1.to_csv('p63enh_starrseq_log2fc_matched_MCF10A_WTp53KO_5variants.csv')\n",
    "df_wide2.to_csv('p63enh_starrseq_log2fc_matched_MCF10Ap53KO_HaCaT_SCC25_WTmut.csv')\n",
    "df_wide3.to_csv('p63enh_starrseq_log2fc_matched_MCF10AGus_TAp63B_WTmut.csv')\n",
    "df_norm_long_3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "WT     25698\n",
       "mut    25698\n",
       "Name: enhancer_variant, dtype: int64"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_norm_long_2.enhancer_variant.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_wide1.to_csv('p63enh_starrseq_log2fc_matched_MCF10A_WTp53KO_5variants.csv')\n",
    "df_wide2.to_csv('p63enh_starrseq_log2fc_matched_MCF10Ap53KO_HaCaT_SCC25_WTmut.csv')\n",
    "df_wide3.to_csv('p63enh_starrseq_log2fc_matched_MCF10AGus_TAp63B_WTmut.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Prepare a table that has GC skew % accross the enhancer for a selected window size (10 nt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "#use enhancers from matched MCf10A WT and p53KO table since it will beused to plot data in Fig. 4\n",
    "df_pko = pd.read_csv('p63enh_starrseq_normCounts_matched_MCF10A_WTp53KO_5variants.csv', usecols=['loc','enhancer_variant','p63RE_type','cell_line','activity'])\n",
    "df_pko = df_pko[(df_pko.cell_line == 'MCF10A p53KO') & (df_pko.enhancer_variant == 'WT')]\n",
    "l = df_pko['loc'].tolist()\n",
    "df_seqs = pd.read_csv('p63enh_starrseq_info_long.csv', usecols=['loc','seq','enhancer_variant'])\n",
    "#we only need seqs from the WT and those that are matched in WT and p53KO\n",
    "df_seqs = df_seqs[(df_seqs.enhancer_variant == 'WT') & (df_seqs['loc'].isin(l))]\n",
    "#make loc:seq dictionary\n",
    "enh = pd.Series(df_seqs.seq.values,index=df_seqs['loc']).to_dict()\n",
    "#set rolling window size & function\n",
    "window = 10\n",
    "def gc_count(region):\n",
    "\tgc = ((region.count('G') + region.count('C')) / window) * 100\n",
    "\treturn gc\n",
    "d={}\n",
    "for loc, seq in enh.items():\n",
    "\ts = seq.upper()\n",
    "\tpos=0\n",
    "\td.update({loc:[]})\n",
    "\twhile pos + window < len(s) + 1:\n",
    "\t\tskew = gc_count(s[pos:pos + window])\n",
    "\t\td[loc].append(skew)\n",
    "\t\tpos += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "#create a new df to add gc skew values for each enhancer sequence\n",
    "df = pd.DataFrame()\n",
    "for loc in d:\n",
    "\ttemp = pd.DataFrame({'GC_skew': d[loc]})\n",
    "\ttemp['loc'] = loc\n",
    "\ttemp.reset_index(inplace=True)\n",
    "\ttemp.rename(columns={'index':'position'}, inplace=True)\n",
    "\tdf = pd.concat([df,temp], ignore_index=True)\n",
    "#add 1 to offset python counting\n",
    "df.position = df.position + 1\n",
    "df = df.merge(df_pko, on='loc', how='left')\n",
    "df.to_csv('p63enh_starrseq_gcskew_matched_MCF10Ap53KO.csv')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
